File: javascript/easy/count_and_say.js
// 38. Count and Say
// https://leetcode.com/problems/count-and-say/

const countAndSay = function(n) {
    var str = '1';
    while (n > 1) {
        var newStr = '', count = 0, say = str[0];
        for (var i = 0; i <str.length; i++) {
            if (str[i] === say) {
                count++;
            } else {
                newStr += count + say;
                count = 1;
                say = str[i];
            }
        }
        str = newStr + count + say;
        n--;
    }
    return str;
};

File: python/easy/count_and_say.py
def countAndSay(self, n):

    if n == 1:
        return "1"

    result = ""
    count = 1
    previous = self.countAndSay(n-1)
    for i in range(len(previous)):
        if i == len(previous) - 1 or previous[i] != previous[i+1]:
            result += str(count) + previous[i]
            count = 1
        else:
            count += 1
    return result


n = 4
countAndSay(n)